llvm/lib/Target/AMDGPU/SIInstrInfo.cpp

```diff
@@ -213,17 +213,20 @@ bool SIInstrInfo::isSafeToSink(MachineInstr &MI,
       // Check if there is a FromCycle that contains SgprDef's basic block but
       // does not contain SuccToSinkTo and also has divergent exit condition.
       while (FromCycle && !FromCycle->contains(ToCycle)) {
-        // After structurize-cfg, there should be exactly one cycle exit.
+        // After structurize-cfg, there should be exactly one cycle exit. Also,
+        // cycle exit block should have exactly one predecessor, the cycle exit.
+        // Early-tailduplication can removed that block so we have to search for
+        // predecessor that is in cycle.
```

Comment:
The comment here is not useful anymore. The logic should work in general CFG. Meanwhile, I would like to see the code cleaner. Instead of getting exit blocks and check its predecessors. Let's just check against cycle exiting blocks.
```
FromCycle->getExitingBlocks(ExitingBlocks);
for (auto *Exiting : ExistingBlocks)
  if (hasDivergentBranch(Exiting))
    return false;
```

Reply:
`LoopInfo` has `getExitingBlocks` but `CycleInfo` does not.

I would be happy with removing the comment but @ssahasra suggested to keep it.

Reply:
I like getExitingBlocks more.
btw, is that the formal name for blocks that contain branch that exits the cycle?

Reply:
Should I switch to using LoopInfo then? For our targets cycles at this point should all be natural loops.

Reply:
I maintain it is fully invalid to expect anything out of the MIR CFG. If it isn't enforced by the machine verifier, it isn't required. You cannot assume natural loops